scvi/external/cell2location/_base.py

```diff
@@ -0,0 +1,1060 @@
+        )
+
+        iter_iterator = tqdm(range(max_epochs))
+        hist = []
+        for it in iter_iterator:
+
+            loss = svi.step(*args, **kwargs)
+            iter_iterator.set_description(
+                "Epoch " + "{:d}".format(it) + ", -ELBO: " + "{:.4e}".format(loss)
+            )
+            hist.append(loss)
+
+            if it % 500 == 0:
+                torch.cuda.empty_cache()
+
+        if continue_training and self.is_trained_:
+            # add ELBO listory
+            hist = self.module.history_ + hist
+        self.module.history_ = hist
+        self.module.is_trained_ = True
+        self.history_ = hist
+        self.is_trained_ = True
+
+    def _train_minibatch(
+        self,
+        max_epochs: Optional[int] = None,
+        max_steps: Optional[int] = None,
+        use_gpu: bool = False,
+        plan_kwargs: Optional[dict] = None,
+        trainer_kwargs: Optional[dict] = None,
+        lr: float = 0.01,
+        autoencoding_lr: Optional[float] = None,
+        clip_norm: float = 200,
+        early_stopping: bool = False,
+        continue_training: bool = True,
+    ):
+        """
+        Private method for training using minibatches (scVI interface and pytorch lightning).
+
+        Parameters
+        ----------
+        max_epochs
+            Number of training epochs / iterations
+        max_steps
+            Number of training steps
+        use_gpu
+            Bool, use gpu?
+        plan_kwargs
+            Training plan arguments such as optim and loss_fn
+        trainer_kwargs
+            Arguments for scvi.train.Trainer.
+        early_stopping
+            Bool, use early stopping? (not tested)
+        continue_training
+            When the model is already trained, should calling .train() continue training? (False = restart training)
+
+        Returns
+        -------
+        ELBO history in self.module.history_
+
+        """
+
+        if not continue_training or not self.is_trained_:
+            # models share param store, make sure it is cleared before training
+            pyro.clear_param_store()
+
+        gpus, device = parse_use_gpu_arg(use_gpu)
+        if max_epochs is None:
+            n_obs = self.adata.n_obs
+            max_epochs = np.min([round((20000 / n_obs) * 400), 400])
+
+        plan_kwargs = plan_kwargs if isinstance(plan_kwargs, dict) else dict()
+        trainer_kwargs = trainer_kwargs if isinstance(trainer_kwargs, dict) else dict()
+
+        batch_size = self.module.model.batch_size
+        # initialise guide params (warmup)
+        train_dl = AnnDataLoader(self.adata, shuffle=True, batch_size=batch_size)
+        for tensor_dict in train_dl:
+            args, kwargs = self.module._get_fn_args_from_batch(tensor_dict)
+            args = [a.to(device) for a in args]
+            kwargs = {k: v.to(device) for k, v in kwargs.items()}
+            self.to_device(device)
+            self.module.guide(*args, **kwargs)
+            break
```

Comment:
for this, we can use the callback in the test_pyro file (I think we can move it to the `scvi.train` module in the callbacks file)